### PR TITLE
Unify refueling and routing code

### PR DESCRIPTION
**Pull request by tehhowch**
- Extends advanced refueling logic to ships moving independently & following orders
 - Extends advanced refueling logic to in-system escorts
 - Chooses the best route for a ship to reach its destination system

@flaviojs please check this pull for differences in escort behavior compared to your pull - I'm pretty sure it keeps all the behaviors you added while extending them to MoveIndependent (where they can serve `personality uninterested` and #2990 ).

### CODE REVIEW

File: source/AI.cpp [hunk 1]
```diff
@@ -103,6 +103,109 @@ namespace {
...
+		return false;
+	}
+	// Wrapper for ship-system uses.
+	bool ShouldRefuel(const Ship &ship, const System *to)
+	{
+		if(!to || ship.Fuel() == 1. || !ship.GetSystem()->HasFuelFor(ship))
+			return false;
+		double needed = ship.JumpFuel(to);
+		if(needed && to->HasFuelFor(ship))
+			return ship.Fuel() * ship.Attributes().Get("fuel capacity") < needed;
```
Comment (flaviojs) [inline source/AI.cpp]: When there is no fuel capacity, it tells the ship to refuel.

File: source/AI.cpp [hunk 2]
```diff
@@ -1105,13 +1213,13 @@ void AI::MoveIndependent(Ship &ship, Command &command) const
 		vector<int> systemWeights;
 		int totalWeight = 0;
 		const set<const System *> &links = ship.Attributes().Get("jump drive")
-			? ship.GetSystem()->Neighbors() : ship.GetSystem()->Links();
+			? origin->Neighbors() : origin->Links();
```
Comment (flaviojs) [inline source/AI.cpp]: Hmm, looks like ships with a jump drive and a hyperdrive never go through links that are too far to be neighbors.

Comment (tehhowch) [reply source/AI.cpp]: Neighbors() includes all Links()

Comment (flaviojs) [reply source/AI.cpp]: Hmm, then ships that only have a jump drive can travel farther that the jump distance when a long hyperdrive link is available.

Comment (tehhowch) [reply source/AI.cpp]: That is explicitly intended behavior

Comment (flaviojs) [reply source/AI.cpp]: Ok.

File: source/AI.cpp (hunk 1 above)
Comment (tehhowch) [reply source/AI.cpp]: Should be addressed now.

**Comment by flaviojs** [comment]
There is a case that can be improved further. When my escort is jumping with me, it could consider the next refuel stop of my travel plan.

I found this while playing as a merchant with the initial Star Barge as escort and the travel plan Wazn -> Alhena (fuel) -> Menkalinan (no fuel) -> Capella (restricted) -> Procyon (no fuel) -> Sirius.
It get's stuck in Capella and then Procyon. If it had refueled in Alhena it would only be stuck in Procyon.

**Comment by tehhowch** [comment]
@flaviojs that happens on both `master` and with this PR?

Also, if you know the changes you'd make, I'm happy to merge them in. Just open the PR to my branch

**Comment by flaviojs** [comment]
After rechecking, the example behavior is for ```master```.

In this PR it goes refuel when it doesn't really need to, which isn't that bad.
With a Mule as escort and the travel plan Epsilon Leonis -> Gomeisa (fuel) -> Zosma (fuel) -> Algieba (no fuel) -> Tania Australis (fuel), it made an unnecessary refuel at Zosma.

I'll go see what would need to change.

**Comment by flaviojs** [comment]
I made a PR to your branch tehhowch/endless-sky#27.

A better approach would be to make a new class TravelPlan that supported multiple destinations and filled in the gaps with DistanceMap as needed. It would be able to handle link changes and unexpected system changes (by landing on a wormhole) with the same "fill in the gaps" approach.

**Pull request merged**